x/emissions/keeper/inference_synthesis/network_inference_builder.go

```diff
@@ -1,215 +1,352 @@
+			Forecasters:          args.Forecasters,
+			ForecasterToForecast: args.ForecasterToForecast,
+			ForecasterToRegret:   args.ForecasterToRegret,
+			NetworkCombinedLoss:  args.NetworkCombinedLoss,
+			EpsilonTopic:         args.EpsilonTopic,
+			PNorm:                args.PNorm,
+			CNorm:                args.CNorm,
+		},
+	)
 	if err != nil {
-		return alloraMath.Dec{}, errorsmod.Wrapf(err, "Error recalculating forecast-implied inferences")
+		return alloraMath.Dec{}, errorsmod.Wrapf(err, "calcOneOutInfererInference() error recalculating forecast-implied inferences")
 	}
 
-	paletteCopy.ForecastImpliedInferenceByWorker = palette.ForecastImpliedInferenceByWorker
-
-	// Get regrets for the remaining inferers
-	paletteCopy.InfererRegrets = make(map[string]*alloraMath.Dec)
-	for _, inferer := range totalInferers {
-		regret, _, err := paletteCopy.K.GetOneOutInfererInfererNetworkRegret(paletteCopy.Ctx, paletteCopy.TopicId, withheldInferer, inferer)
-		if err != nil {
-			return alloraMath.Dec{}, errorsmod.Wrapf(err, "Error getting one-out inferer regret")
-		}
-		paletteCopy.InfererRegrets[inferer] = &regret.Value
-	}
 	// Get regrets for the forecasters
-	paletteCopy.ForecasterRegrets = make(map[string]*alloraMath.Dec)
-	for _, forecaster := range paletteCopy.Forecasters {
-		regret, _, err := paletteCopy.K.GetOneOutInfererForecasterNetworkRegret(paletteCopy.Ctx, paletteCopy.TopicId, withheldInferer, forecaster)
+	remainingForecasterRegrets := make(map[string]*alloraMath.Dec)
+	for _, forecaster := range args.Forecasters {
+		regret, _, err := args.K.GetOneOutInfererForecasterNetworkRegret(args.Ctx, args.TopicId, args.WithheldInferer, forecaster)
 		if err != nil {
-			return alloraMath.Dec{}, errorsmod.Wrapf(err, "Error getting one-out forecaster regret")
+			return alloraMath.Dec{}, errorsmod.Wrapf(err, "calcOneOutInfererInference() error getting one-out forecaster regret")
 		}
-		paletteCopy.ForecasterRegrets[forecaster] = &regret.Value
-	}
-
-	weights, err := paletteCopy.CalcWeightsGivenWorkers()
+		remainingForecasterRegrets[forecaster] = &regret.Value
+	}
+
+	weights, err := calcWeightsGivenWorkers(
+		args.Logger,
+		remainingInferers,
+		args.Forecasters,
+		remainingInfererRegrets,
+		remainingForecasterRegrets,
+		args.EpsilonTopic,
+		args.PNorm,
+		args.CNorm,
+	)
 	if err != nil {
-		return alloraMath.Dec{}, errorsmod.Wrapf(err, "Error calculating one-out inference for forecaster")
-	}
-
-	oneOutNetworkInferenceWithoutInferer, err := paletteCopy.CalcWeightedInference(weights)
+		return alloraMath.Dec{}, errorsmod.Wrapf(err, "calcOneOutInfererInference() error calculating one-out inference for forecaster")
+	}
+
+	oneOutNetworkInferenceWithoutInferer, err = calcWeightedInference(
+		args.Logger,
+		args.AllInferersAreNew,
+		remainingInferers,
+		remainingInfererToInference,
+		remainingInfererRegrets,
+		args.Forecasters,
+		remainingForecasterRegrets,
+		forecasterToForecastImpliedInference,
+		weights,
+		args.EpsilonSafeDiv,
+	)
```

Comment:
if num args >3 or 4, plz write a struct. That is truly Option A

Reply:
missed this spot, thought I had gotten everything, will address